lib/remoteproc/remoteproc_virtio.c

```diff
@@ -35,13 +35,168 @@
 
 #include <openamp/remoteproc.h>
 #include <openamp/remoteproc_virtio.h>
+#include <openamp/virtqueue.h>
 
-static int rproc_virtio_notify(unsigned int notifyid, struct remoteproc *rproc)
+
+static void rproc_virtio_virtqueue_notify(struct virtqueue *vq)
+{
+	struct virtio_device *vdev;
+	struct remoteproc_virtio *rpvdev;
+	struct remoteproc_vring *rvring;
+	unsigned int vq_id = vq->vq_queue_index;
+
+	vdev = vq->vq_dev;
+	rpvdev = metal_container_of(vdev, struct remoteproc_virtio, vdev);
+	metal_assert(vq_id <= vdev->num_vrings);
+	rvring = rpvdev->rvrings[vq_id];
+	rpvdev->notify(rpvdev->priv, rvring->notifyid);
+}
+
+static int rproc_virtio_create_virtqueues(struct virtio_device *vdev,
+					  unsigned int flags,
+					  unsigned int nvqs,
+					  const char *names[],
+					  vq_callback *callbacks[],
+					  struct virtqueue *vqs[])
+{
+	struct remoteproc_virtio *rpvdev;
+	struct remoteproc_vring *rvring;
+	struct virtio_device *vdev;
+	struct vring_alloc_info vring_info;
+	unsigned int num_vrings, i;
+	int ret;
+
+	(void)flags;
+	rpvdev = metal_container_of(vdev, struct remoteproc_virtio, vdev);
+	num_vrings = rpvdev->vdev.num_vrings;
+	if (nvqs > num_vrings)
+		return -EINVAL;
+	vdev = &rpvdev->vdev;
+	/* Initialize virtqueue for each vring */
+	for (i = 0; i < nvqs; i++) {
+		rvring = &rpvdev->vrings[i];
+		if (vdev->role == VIRTIO_DEV_GUEST) {
+			size_t offset;
+			struct metal_io_region *io = rvring->io;
+			unsigned int num_descs = rvring->num_descs;
+			unsigned int align = rvring->align;
+
+			vring_info.vaddr = rvring->va;
+			vring_info.align = rvring->align;
+			vring_info.num_descs = rvring->num_descs;
+			offset = metal_io_virt_to_offset(io, rvring->va);
+			metal_io_block_set(io, offset, 0,
+					   vring_size(num_descs, align));
+			ret = virtqueue_create(vdev, i,
```

Comment:
virtqueues have to be created for both,  host and guest, only the block set should be done  for "host"